gnovm/pkg/gnolang/preprocess.go

```diff
@@ -1872,7 +1872,93 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				} else { // ASSIGN, or assignment operation (+=, -=, <<=, etc.)
 					// NOTE: Keep in sync with DEFINE above.
 					if len(n.Lhs) > len(n.Rhs) {
-						// check is done in assertCompatible
+						// check is done in assertCompatible where we also
+						// asserted we have at lease one element in Rhs
+						if cx, ok := n.Rhs[0].(*CallExpr); ok {
+							// we decompose the a,b = x(...) for named and unamed
+							// type value return in an assignments
+							// Call case: a, b = x(...)
+							ift := evalStaticTypeOf(store, last, cx.Func)
+							cft := getGnoFuncTypeOf(store, ift)
+							if len(n.Lhs) != len(cft.Results) {
+								panic(fmt.Sprintf(
+									"assignment mismatch: "+
+										"%d variables but %s returns %d values",
+									len(n.Lhs), cx.Func.String(), len(cft.Results)))
+							}
+
+							// check if we we need to decompose for named typed conversion in the function return results
+							var decompose bool
+
+							for i, rhsType := range cft.Results {
+								lt := evalStaticTypeOf(store, last, n.Lhs[i])
+								if lt != nil && isNamedConversion(rhsType.Type, lt) {
+									decompose = true
+									break
+								}
+							}
+							if decompose {
+								// only enter this section if cft.Results to be converted to Lhs type for named type conversion.
+								// decompose a,b = x()
+								// .tmp1, .tmp2 := x()  assignment statement expression (Op=DEFINE)
+								// a,b = .tmp1, .tmp2   assignment statement expression ( Op=ASSIGN )
+								// add the new statement to last.Body
+
+								// step1:
+								// create a hidden var with leading . (dot) the curBodyLen increase every time when there is an decompostion
+								// because there could be multiple decomposition happens
+								// we use both stmt index and resturn result number to differentiate the .tmp variables created in each assignment decompostion
+								// ex. .tmp_3_2: this variable is created as the 3rd statement in the block, the 2nd parameter returned from x(),
+								// create .tmp_1_1, .tmp_1_2 .... based on number of result from x()
+								var tmpExprs Exprs
+								for i := range cft.Results {
+									rn := fmt.Sprintf(".tmp_%d_%d", index, i)
+									tmpExprs = append(tmpExprs, Nx(rn))
+								}
+								// step2:
+								// .tmp1, .tmp2 := x()
+								dsx := &AssignStmt{
+									Lhs: tmpExprs,
+									Op:  DEFINE,
+									Rhs: n.Rhs,
+								}
+								dsx.SetLine(n.Line)
+								dsx = Preprocess(store, last, dsx).(*AssignStmt)
+
+								// step3:
+
+								// a,b = .tmp1, .tmp2
+								// assign stmt expression
+								// the right hand side will be converted to  call expr for nameed/unnamed covnersion
+								// we make a copy of tmpExprs to prevent dsx.Lhs in the preview statement changing by the side effect
```

Comment:
```suggestion
// the right hand side will be converted to call expr for named/unnamed conversion,
// we make a copy of tmpExprs to prevent dsx.Lhs in the previous statement 
changing by the side effect
```